### PR TITLE
Add opt-in support for datadog client

### PR DESCRIPTION
**Pull request by MatthewSBarnes**
@markstory 

We want to support datadog so we can use tags. 
Instead of using the datadog lib specifically, for each of our services, this change makes it easier for other apps to switch over to datadog.

### CODE REVIEW

**Comment by markstory** [comment]
> We want to support datadog so we can use tags.

So how are you going to use tags?

The current changes look great, but I don't see any support for tags in any of the methods.

**Comment by markstory** [comment]
Given that we'd be supporting two vendors, we should also extend the test suite so that both vendors are tested. Right now we won't have any confidence that all the sugar methods work with both statsd and datadog.

File: tests/test_functions.py [hunk 1]
```diff
@@ -1,69 +1,103 @@
...
+        client = statsdecor.client()
+        assert client.port == 9999, 'port should match'
+
+    def test_configure_and_create__with_prefix(self):
+        statsdecor.configure(prefix='statsdecor')
+        client = statsdecor.client()
+        assert client.namespace == 'statsdecor', 'namespace should match prefix'
+
+    def test_configure_and_create__with_fields_not_in_whitelist(self):
+        statsdecor.configure(maxudpsize=512)
```
Comment (iamjillsanluis) [inline tests/test_functions.py]: 👍

**Comment by skaasten** [comment]
Seems reasonable to me 👍

File: tests/test_decorators.py [hunk 2]
```diff
@@ -1,40 +1,90 @@
...
 
-def test_timed():
-    @decorators.timed('a.metric')
-    def test_fn(*args, **kwargs):
-        assert_arguments(args, kwargs)
+class TestStatsdDefaultClient(BaseDecoratorTestCase):
+    def setup(self):
+        self.tags = ['StatsClient_doesnt_do_tags!']
+        self.client_class = StatsdClient
+        statsdecor.configure()
```
[the remark this answers is not in the document]
Comment (MatthewSBarnes) [reply tests/test_decorators.py]: These tests just assert that the decorators actually call `stats.incr(...)`, `stats.decr(...)`, and `stats.timer(...)` on L19, L40, L59, respectively.

Should I take it a level deeper?

File: tests/test_functions.py [hunk 3]
```diff
@@ -1,69 +1,103 @@
...
+
+class TestStatsdDefaultClient(BaseFunctionTestCase):
+    def setup(self):
+        self.client_class = StatsdClient
+        statsdecor.configure(vendor='')
+
+    def test_client_error_no_config(self, monkeypatch):
+        # TODO: DIG UP HISTORY what is the below referring to ?
+        # When the client hasn't been initialized,
+        # an exception should be raised by client()
```
[the remark this answers is not in the document]
Comment (MatthewSBarnes) [reply tests/test_functions.py]: Ah so the test is a misnomer, cool. 
I'll drop the comment and update the test name.

File: statsdecor/clients.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+        self._assert_no_tags(tags)
+        super(StatsdClient, self).timing(stat=name, delta=value, rate=rate)
+
+    def timer(self, name, tags=None):
+        self._assert_no_tags(tags)
+        super(StatsdClient, self).timer(stat=name)
+
+    def _assert_no_tags(self, tags):
+        if tags:
+            raise ValueError('Tagging is not supported by StatsdClient.\n tags: {}'.format(tags))
```
[the remark this answers is not in the document]
Comment (MatthewSBarnes) [reply statsdecor/clients.py]: I'll update it, thank you 👍

**Pull request merged**